### PR TITLE
Multiprocessing support for off policy algorithms

### LINKED ISSUE #179: [Feature request] Adding multiprocessing support for off policy algorithms
I am in the process of adding multiprocessing(vectorized envs) support for off-policy algorithms (TD3, SAC, DDPG etc), I've added support for sampling multiple actions and updates timesteps appropriately to the number of vectorized environments. The modified code can run without throwing an error, but the algorithms don't really converge anymore. 
I tried on OpenAI Gym's Pendulum-v0, where single instance envs made from `make_vec_env('Pendulum-v0', n_envs = 1, vec_env_cls=DummyVecEnv)` trains fine.  If I specify multiple instances such as `make_vec_env('Pendulum-v0', n_envs = 2, vec_env_cls=DummyVecEnv)` or `make_vec_env('Pendulum-v0', n_envs = 2, vec_env_cls=SubprocVecEnv)`, then the algorithms don't converge at all. 
Here's a warning message that I get, which I suspect is closely related to the non-convergence.
 ```
/home/me/code/stable-baselines3/stable_baselines3/sac/sac.py:237: UserWarning: Using a target size (torch.Size([256, 2])) that is different
 to the input size (torch.Size([256, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.                                                                                                                                             critic_loss = 0.5 * sum([F.mse_loss(current_q, q_backup) for current_q in current_q_estimates])   
```
It appears to me that the replay buffer wasn't not retrieving `n_envs` samples thus the loss target had to rely on broadcasting 

Some pointers on modifying the replay buffer so it would support multiprocessing would be much appreciated! If the authors would like, I can create a PR

https://github.com/yonkshi/stable-baselines3/commit/157971357a28be8435d09cfccec1d4258b220a6a

### PR DESCRIPTION
**Pull request by araffin**
## Description

closes #179

Add support for multiprocessing, only `HerReplayBuffer` not supported (done in #654 ).

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by araffin** [comment]
@yonkshi @SonSang could you also take a look and review this PR?

**Comment by Miffyli** [comment]
Would you have any performance metrics to share vs. single-threaded (agent performance, training time)? I will try to take a look before end of the week :)

**Comment by araffin** [comment]
so I've got something qualitative so you get the general idea of the pros/cons of multi envs.
This is SAC on Pendulum on cpu with (in order):
- one env
- two envs
- four envs
- height envs
- four envs and `gradient_steps=2` (instead of 1)

Sample efficiency:
Training_Episodic_Reward_steps

Wall clock time

Training_Episodic_Reward

more envs = less sample efficient and faster wall clock time but you can retrieve some sample efficiency by having more gradients steps , in that case you will lose some speed (especially in that experiment with my laptop on cpu only)

File: tests/test_spaces.py [hunk 1]
```diff
@@ -69,8 +69,13 @@ def test_action_spaces(model_class, env):
...
+@pytest.mark.parametrize("model_class", [A2C, PPO, DQN])
 @pytest.mark.parametrize("env", ["Taxi-v3"])
 def test_discrete_obs_space(model_class, env):
     env = make_vec_env(env, n_envs=2, seed=0)
-    model_class("MlpPolicy", env, n_steps=256).learn(500)
+    kwargs = {}
+    if model_class == DQN:
+        kwargs = dict(buffer_size=1000, learning_starts=100)
+    else:
+        kwargs = dict(n_steps=256)
```
Comment (Miffyli) [inline tests/test_spaces.py:80]: This should probably be lower than training length so that at least one proper update period happens (it probably does, but would be more comfortable to keep the original 500 or so :) )

File: docs/guide/examples.rst [hunk 2]
```diff
@@ -158,6 +158,22 @@ Multiprocessing: Unleashing the Power of Vectorized Environments
           env.render()
 
 
+Multiprocessing with off-policy algorithms
```
Comment (Miffyli) [inline docs/guide/examples.rst:161]: A note about the potential "worse sample efficiency" etc would be nice here. I expect people try increasing number of envs but observe no big speedups in training (depending on the env). A link to the PR commente/experiment?

File: stable_baselines3/common/type_aliases.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ class DictReplayBufferSamples(ReplayBufferSamples):
 
 
 class RolloutReturn(NamedTuple):
-    episode_reward: float
+    episode_reward: float  # not used anymore, kept for backward compat
```
Comment (Miffyli) [inline stable_baselines3/common/type_aliases.py:59]: Is this required by any of our code? I think it would be clearer to outright remove it not to introduce hidden bugs for those who use this variable.

File: stable_baselines3/common/off_policy_algorithm.py [hunk 4]
```diff
@@ -487,20 +493,29 @@ def _store_transition(
...
         # As the VecEnv resets automatically, new_obs is already the
         # first observation of the next episode
-        if done and infos[0].get("terminal_observation") is not None:
-            next_obs = infos[0]["terminal_observation"]
-            # VecNormalize normalizes the terminal observation
-            if self._vec_normalize_env is not None:
-                next_obs = self._vec_normalize_env.unnormalize_obs(next_obs)
-        else:
-            next_obs = new_obs_
+        next_obs = deepcopy(new_obs_)  # use deepcopy for dict observations
```
Comment (Miffyli) [inline stable_baselines3/common/off_policy_algorithm.py:496]: `deepcopy` is only necessary for dict obs, yes? Might be worth saving the processing time by moving it inside the dict-obs check or so.

File: stable_baselines3/common/off_policy_algorithm.py [hunk 5]
```diff
@@ -541,79 +556,72 @@ def collect_rollouts(
...
+                # Sample a new noise matrix
+                self.actor.reset_noise(env.num_envs)
 
-                # Give access to local variables
-                callback.update_locals(locals())
-                # Only stop training if return value is False, not when it is None.
-                if callback.on_step() is False:
-                    return RolloutReturn(0.0, num_collected_steps, num_collected_episodes, continue_training=False)
+            # Select action randomly or according to policy
+            action, buffer_action = self._sample_action(learning_starts, action_noise, env.num_envs)
```
Comment (Miffyli) [inline stable_baselines3/common/off_policy_algorithm.py:583]: Nitpick, but for consistency with `OnPolicyAlgorithm` and for clarity, I would prefer plural forms for variable names (e.g., "actions", "rewards").

File: tests/test_spaces.py (hunk 1 above)
Comment (araffin) [reply tests/test_spaces.py:80]: it does and the 500 was not so good as A2C/PPO would do 512 steps anyway.

Comment (araffin) [reply tests/test_spaces.py:80]: it does and the 500 was not so good as A2C/PPO would do 512 steps anyway.

File: stable_baselines3/common/type_aliases.py (hunk 3 above)
Comment (araffin) [reply stable_baselines3/common/type_aliases.py:59]: I was mainly thinking of people who created custom algorithms, this variable is only meant for logging...
But you are probably right, we should probably just get rid of it.

File: stable_baselines3/common/off_policy_algorithm.py (hunk 4 above)
Comment (Miffyli) [reply stable_baselines3/common/off_policy_algorithm.py:496]: Yup. This sounds a lot like premature optimization, but to my experience it is the copies that bottleneck performance, so might be worth saving up here :)

**Review by Miffyli — APPROVED** [review]
LGTM but need to fix the typing thing :)

**Comment by araffin** [comment]
I reintroduced the `next_obs = deepcopy(new_obs_)` because it prevents many errors (and not only with dict observations)
